var/spack/repos/builtin/packages/python/package.py

```diff
@@ -946,6 +854,56 @@ def setup_dependent_build_environment(self, env, dependent_spec):
         pythonpath = ':'.join(python_paths)
         env.set('PYTHONPATH', pythonpath)
 
+        # We need to make sure that the extensions are compiled and linked with
+        # the Spack wrapper. Paths to the executables that are used for these
+        # operations are normally taken from the sysconfigdata file, which we
+        # modify after the installation (see method filter compilers). The
+        # modified file contains paths to the real compilers, not the wrappers.
+        # The values in the file, however, can be overridden with environment
+        # variables. The first variable, CC (CXX), which is used for
+        # compilation, is set by Spack for the dependent package by default.
+        # That is not 100% correct because the value for CC (CXX) in the
+        # sysconfigdata file often contains additional compiler flags (e.g.
+        # -pthread), which we lose by simply setting CC (CXX) to the path to the
+        # Spack wrapper. Moreover, the user might try to build an extension with
+        # a compiler that is different from the one that was used to build
+        # Python itself, which might have unexpected side effects. However, the
+        # experience shows that none of the above is a real issue and we will
+        # not try to change the default behaviour. Given that, we will simply
+        # try to modify LDSHARED (LDCXXSHARED), the second variable, which is
+        # used for linking, in a consistent manner.
+
+        for compile_var, link_var in [('CC', 'LDSHARED'),
+                                      ('CXX', 'LDCXXSHARED')]:
+            # First, we get the values from the sysconfigdata:
+            config_compile = self.get_config_var(compile_var)
+            config_link = self.get_config_var(link_var)
+
+            # The dependent environment will have the compilation command set to
+            # the following:
+            new_compile = join_path(
+                spack.paths.build_env_path,
+                dependent_spec.package.compiler.link_paths[compile_var.lower()])
+
+            # Normally, the link command starts with the compilation command:
+            if config_link.startswith(config_compile):
+                new_link = new_compile + config_link[len(config_compile):]
+            else:
+                # with a few exceptions when the compilation command is either
+                # not mentioned in the link command at all (ld is used directly
+                # instead, which should be fine since the first ld in the PATH
+                # is the Spack wrapper) or somewhere in the middle surrounded
+                # with spaces:
```

Comment:
This PR looks good to me but one thing it took me a while to understand is that this may not affect the linking command at all (e.g. if the `.replace` has no effect because it doesn't find the search string). I'll suggest the following to make that clearer and to also to include https://github.com/spack/spack/pull/21149#discussion_r561114580 which I thought was helpful:

```
# We try to replace the compiler command if it appears "in the middle" of the
# link command; to avoid mistaking some substring of a path for the compiler
# (e.g. to avoid replacing "gcc" in "-L/path/to/gcc/"), we require that the compiler
# command be surrounded by spaces
# Note this may leave "config_link" unchanged if the compilation command does
# not appear in the link command at all, for example if "ld" is invoked directly
# (no change would be required in that case because Spack arranges for the
# Spack ld wrapper to be the first instance of "ld" in PATH). 
```

If you agree with that wording, could you update the comment?